### PR TITLE
Making payload field optional for the v1 and v2 message endpoint [RHCLOUD-20916]

### PR DESCRIPTION
**Pull request by tahmidefaz**
## What?
The `payload` field for the post body of the v1 and v2 message endpoint is now optional.

[Jira ticket](https://issues.redhat.com/browse/RHCLOUD-20916)

## Why?
playbook-dispatcher's cancel operation needs to pass a `directive` that does not have a `payload` associated with it

## How?
Removed the `validate:"required"` json tags for the Payload fields.

## Testing
Tests added for both the v1 and v2 to demonstrate the behavior.

## Anything Else?
Any other notes about the PR that would be useful for the reviewer. 

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: internal/controller/api/connection_mediator_v2_test.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	. "github.com/onsi/gomega"
+
+	"github.com/RedHatInsights/cloud-connector/internal/config"
+	"github.com/RedHatInsights/cloud-connector/internal/domain"
+
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+)
+
+const (
```
Comment (dehort) [inline internal/controller/api/connection_mediator_v2_test.go:20]: Let's try to get rid of these constants.  I did stuff like this in the old tests, but I think we should start cleaning those up and trying to make the tests more self contained.

File: internal/controller/api/connection_mediator_v2_test.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	MESSAGE_ENDPOINT_V2 = URL_BASE_PATH + "/v2/connections/345/message"
+	ACCOUNT_NUMBER      = domain.AccountID("1234")
+)
+
+func MockedGetConnectionByClientID(context.Context, *logrus.Entry, domain.OrgID, domain.ClientID) (domain.ConnectorClientState, error) {
```
Comment (dehort) [inline internal/controller/api/connection_mediator_v2_test.go:25]: What do you think about modifying this so that we can pass in the expected org-id, account and client-id?  I think that would be a cleaner approach and would likely be useful moving forward.  One way to approach this would be the following:

```go
func mockedGetConnectionByClientID(expectedOrgId domain.OrgID, expectedAccount domain.AccountID, expectedClientId domain.ClientID) connection_repository.GetConnectionByClientID {
	return func(ctx context.Context, log *logrus.Entry, actualOrgId domain.OrgID, actualClientId domain.ClientID) (domain.ConnectorClientState, error) {
		if actualOrgId != expectedOrgId {
			return domain.ConnectorClientState{}, fmt.Errorf("Actual org id does not match expected org id")
		}

		if actualClientId != expectedClientId {
			return domain.ConnectorClientState{}, fmt.Errorf("Actual client id does not match expected client id")
		}

		return domain.ConnectorClientState{OrgID: actualOrgId, Account: expectedAccount, ClientID: actualClientId}, nil
	}
}
```

This allows us to pass in the `expected` values and confirm that those expected values match how the code is called during the test.

It also allows us to make the tests a bit more self-contained (less globals, less hard coded values, etc).

Also notice that I'm not exporting the mock out of the package (little 'm' in mockedGetConnectionByClientID).  Always try to expose a little as possible.

File: internal/controller/api/connection_mediator_v2_test.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	return domain.ConnectorClientState{Account: ACCOUNT_NUMBER, ClientID: "345"}, nil
+}
+
+func MockedGetConnectionsByOrgID(context.Context, *logrus.Entry, domain.OrgID, int, int) (map[domain.ClientID]domain.ConnectorClientState, int, error) {
```
Comment (dehort) [inline internal/controller/api/connection_mediator_v2_test.go:29]: Same as the previous comment

File: internal/controller/api/connection_mediator_v2_test.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	BeforeEach(func() {
+		apiMux := mux.NewRouter()
+		cfg := config.GetConfig()
+
+		proxyFactory := &MockClientProxyFactory{}
+
+		cm = NewConnectionMediatorV2(MockedGetConnectionByClientID, MockedGetConnectionsByOrgID, proxyFactory, apiMux, URL_BASE_PATH, cfg)
+		cm.Routes()
+
+		validIdentityHeader = buildIdentityHeader(ACCOUNT_NUMBER, "Associate")
```
Comment (dehort) [inline internal/controller/api/connection_mediator_v2_test.go:49]: The `buildIdentityHeader()` method really needs to be modified to take in the org-id but we do not have to make that change now.  We should probably make that change in a different pr.

**Review by dehort — CHANGES_REQUESTED** [review]
This is looking very good.  I like how the v2 tests are looking (thank you for putting these together).  I left a couple of comments.

File: internal/controller/api/connection_mediator_v2_test.go (hunk 2 above)
Comment (tahmidefaz) [reply internal/controller/api/connection_mediator_v2_test.go]: Great idea! Updated!

File: internal/controller/api/connection_mediator_v2_test.go (hunk 4 above)
Comment (tahmidefaz) [reply internal/controller/api/connection_mediator_v2_test.go]: Agree, I needed to match the org id I used for the mock functions with the one that's inside `buildIdentityHeader()` :+1:

**Review by dehort — APPROVED** [review]
LGTM!
